### PR TITLE
SONAR-7835 WS api/permissions/users return permissions in response

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionDao.java [hunk 1]
```diff
@@ -49,19 +51,38 @@ public PermissionDao(MyBatis myBatis) {
...
+  public void selectUsersByQuery(DbSession dbSession, PermissionQuery query, ResultHandler handler) {
+    mapper(dbSession).selectUsersByQuery(query, new RowBounds(query.getPageOffset(), query.getPageSize()), handler);
+  }
+
+  public int countUsersByQuery(DbSession dbSession, PermissionQuery query) {
+    return mapper(dbSession).countUsersByQuery(query);
+  }
+
+  public List<UserPermissionDto> selectUserPermissionsByQuery(DbSession dbSession, PermissionQuery query) {
+    if (query.getLogins() != null && query.getLogins().isEmpty()) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionDao.java]: BLOCKER NullPointerException might be thrown as 'getLogins' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* BLOCKER 1 blocker
* INFO 7 info

Watch the comments in this conversation to review them.

#### 7 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. INFO [PermissionTemplateDao.java#L266](https://github.com/SonarSource/sonarqube/blob/e38046d794dc06a8849e8159e7160bf9af775959/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L266): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L293](https://github.com/SonarSource/sonarqube/blob/e38046d794dc06a8849e8159e7160bf9af775959/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L293): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L322](https://github.com/SonarSource/sonarqube/blob/e38046d794dc06a8849e8159e7160bf9af775959/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L322): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L344](https://github.com/SonarSource/sonarqube/blob/e38046d794dc06a8849e8159e7160bf9af775959/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L344): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L372](https://github.com/SonarSource/sonarqube/blob/e38046d794dc06a8849e8159e7160bf9af775959/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L372): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDto.java#L58](https://github.com/SonarSource/sonarqube/blob/e38046d794dc06a8849e8159e7160bf9af775959/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDto.java#L58): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDto.java#L66](https://github.com/SonarSource/sonarqube/blob/e38046d794dc06a8849e8159e7160bf9af775959/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDto.java#L66): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* BLOCKER 1 blocker
* INFO 7 info

Watch the comments in this conversation to review them.

#### 7 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. INFO [PermissionTemplateDao.java#L266](https://github.com/SonarSource/sonarqube/blob/4cc4c6ec700cae5e3e5560dd89c64a7858cf5210/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L266): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L293](https://github.com/SonarSource/sonarqube/blob/4cc4c6ec700cae5e3e5560dd89c64a7858cf5210/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L293): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L322](https://github.com/SonarSource/sonarqube/blob/4cc4c6ec700cae5e3e5560dd89c64a7858cf5210/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L322): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L344](https://github.com/SonarSource/sonarqube/blob/4cc4c6ec700cae5e3e5560dd89c64a7858cf5210/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L344): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L372](https://github.com/SonarSource/sonarqube/blob/4cc4c6ec700cae5e3e5560dd89c64a7858cf5210/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L372): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDto.java#L58](https://github.com/SonarSource/sonarqube/blob/4cc4c6ec700cae5e3e5560dd89c64a7858cf5210/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDto.java#L58): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDto.java#L66](https://github.com/SonarSource/sonarqube/blob/4cc4c6ec700cae5e3e5560dd89c64a7858cf5210/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDto.java#L66): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml [hunk 2]
```diff
@@ -40,6 +40,68 @@
     </where>
   </sql>
 
+  <select id="selectUsersByQuery" parameterType="map" resultType="org.sonar.db.permission.UserRef">
```
Comment (julienlancelot) [inline sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml]: Why don't you use UserMapper ?

File: sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml [hunk 3]
```diff
@@ -3,17 +3,20 @@
 
 <mapper namespace="org.sonar.db.permission.PermissionMapper">
 
+  <!-- TODO delete when not used anymore -->
```
Comment (julienlancelot) [inline sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml]: Will you remove it before pushing to master ?

File: sonar-db/src/test/java/org/sonar/db/permission/UserWithPermissionDaoTest.java [hunk 4]
```diff
@@ -191,15 +202,36 @@ public void select_users_with_query() {
...
+  public void select_user_permissions() {
+    UserDto user = userDb.insertUser(newUserDto().setLogin("user-login"));
+    UserDto anotherUser = userDb.insertUser(newUserDto().setLogin("another-login"));
+    ComponentDto project = componentDb.insertComponent(newProjectDto());
+    permissionDb.addProjectPermissionToUser(UserRole.ADMIN, user.getId(), project.getId());
+    permissionDb.addProjectPermissionToUser(UserRole.ADMIN, anotherUser.getId(), project.getId());
+
+    PermissionQuery.Builder dbQuery = PermissionQuery.builder()
+      .setComponentUuid(project.uuid())
+      .setLogins(singletonList("user-login"))
```
Comment (julienlancelot) [inline sonar-db/src/test/java/org/sonar/db/permission/UserWithPermissionDaoTest.java]: You should add a test with more than one login

**Comment by julienlancelot** [comment]
As discussed, please remove UserRef usage and replace it by the user id.

**Comment by julienlancelot** [comment]
Be careful, a SQL query is not working on PostgreSQL : 
~~~
2016.07.06 14:26:07 ERROR web[o.s.s.w.WebServiceEngine] Fail to process request http://localhost:9000/api/permissions/users2
org.apache.ibatis.exceptions.PersistenceException:
### Error querying database.  Cause: org.postgresql.util.PSQLException: ERROR: subquery in FROM must have an alias
  Indice : For example, FROM (SELECT ...) [AS] foo.
  Position : 26
### The error may exist in org.sonar.db.permission.PermissionMapper
### The error may involve org.sonar.db.permission.PermissionMapper.countUsersByQuery-Inline
### The error occurred while setting parameters
### SQL: select count(1)     from (       select distinct            u.login as login, u.name as name, u.email as email, lower(u.name), u.id                 from users u       left join user_roles ur ON ur.user_id=u.id         left join projects p on ur.resource_id = p.id      WHERE  u.active = true     )
### Cause: org.postgresql.util.PSQLException: ERROR: subquery in FROM must have an alias
  Indice : For example, FROM (SELECT ...) [AS] foo.
  Position : 26
~~~

**Comment by julienlancelot** [comment]
When the 'q' is too small, the http status is well 400, but the error message is empty.

File: sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml (hunk 3 above)
Comment (teryk) [inline sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml]: I remove the comment as I won't remove them before pushing to master

File: sonar-db/src/test/java/org/sonar/db/permission/UserWithPermissionDaoTest.java (hunk 4 above)
Comment (teryk) [inline sonar-db/src/test/java/org/sonar/db/permission/UserWithPermissionDaoTest.java]: ok

File: sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml [hunk 5]
```diff
@@ -45,17 +45,20 @@
...
   </sql>
 
-  <select id="selectUsersByQuery" parameterType="map" resultType="org.sonar.db.permission.UserRef">
-    select distinct <include refid="userColumns" />
-    <include refid="usersByQuery"/>
-    order by lower(u.name), u.name, u.id
+  <select id="selectLoginsByPermissionQuery" parameterType="map" resultType="string">
+    select u.login
+    from (
+      select distinct <include refid="userColumns" />
```
Comment (julienlancelot) [inline sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml]: Why not directly returning only login here ?

Comment (julienlancelot) [inline sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml]: Do you still need to get email (defined in userColumns)

Comment (julienlancelot) [inline sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml]: Last point, you should see with duke in order to :
- either validate that this is the right query
- either improve the query

Comment (teryk) [inline sonar-db/src/main/resources/org/sonar/db/permission/PermissionMapper.xml]: you're right about email, I can remove it but I must keep name and id to be able to order by with a distinct

File: sonar-db/src/main/java/org/sonar/db/user/UserDao.java [hunk 6]
```diff
@@ -102,6 +102,10 @@ public UserDto selectActiveUserByLogin(DbSession session, String login) {
     return executeLargeInputs(logins, session.getMapper(UserMapper.class)::selectByLogins);
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/user/UserDao.java]: MINOR Remove this use of "selectByLogins"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-db/src/main/java/org/sonar/db/user/UserDao.java [hunk 7]
```diff
@@ -102,6 +102,10 @@ public UserDto selectActiveUserByLogin(DbSession session, String login) {
...
   }
 
+  /**
+   * @deprecated since 6.0 please use {@link #selectByLogins(DbSession, Collection)} instead
+   */
+  @Deprecated
   public List<UserDto> selectByLogins(Collection<String> logins) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/user/UserDao.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionDao.java [hunk 8]
```diff
@@ -49,19 +51,38 @@ public PermissionDao(MyBatis myBatis) {
...
+  public List<String> selectLoginsByPermissionQuery(DbSession dbSession, PermissionQuery query) {
+    return mapper(dbSession).selectLoginsByPermissionQuery(query, new RowBounds(query.getPageOffset(), query.getPageSize()));
+  }
+
+  public int countUsersByQuery(DbSession dbSession, PermissionQuery query) {
+    return mapper(dbSession).countUsersByPermissionQuery(query);
+  }
+
+  public List<UserPermissionDto> selectUserPermissionsByQuery(DbSession dbSession, PermissionQuery query) {
+    if (query.getLogins() != null && query.getLogins().isEmpty()) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionDao.java]: BLOCKER NullPointerException might be thrown as 'getLogins' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues
* BLOCKER 1 blocker
* MINOR 1 minor
* INFO 8 info

Watch the comments in this conversation to review them.

#### 7 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. INFO [PermissionTemplateDao.java#L266](https://github.com/SonarSource/sonarqube/blob/e1f6fd6d5fc0baad14120cf120cb15847337744c/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L266): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L293](https://github.com/SonarSource/sonarqube/blob/e1f6fd6d5fc0baad14120cf120cb15847337744c/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L293): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L322](https://github.com/SonarSource/sonarqube/blob/e1f6fd6d5fc0baad14120cf120cb15847337744c/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L322): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L344](https://github.com/SonarSource/sonarqube/blob/e1f6fd6d5fc0baad14120cf120cb15847337744c/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L344): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L372](https://github.com/SonarSource/sonarqube/blob/e1f6fd6d5fc0baad14120cf120cb15847337744c/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L372): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDto.java#L58](https://github.com/SonarSource/sonarqube/blob/e1f6fd6d5fc0baad14120cf120cb15847337744c/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDto.java#L58): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDto.java#L66](https://github.com/SonarSource/sonarqube/blob/e1f6fd6d5fc0baad14120cf120cb15847337744c/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDto.java#L66): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by julienlancelot** [comment]
Please validate the SQL query with duke, otherwise LGTM.

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues
* BLOCKER 1 blocker
* MINOR 1 minor
* INFO 8 info

Watch the comments in this conversation to review them.

#### 7 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. INFO [PermissionTemplateDao.java#L266](https://github.com/SonarSource/sonarqube/blob/eede4bb37ed5cbc255d60d3651ef969d44f6705f/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L266): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L293](https://github.com/SonarSource/sonarqube/blob/eede4bb37ed5cbc255d60d3651ef969d44f6705f/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L293): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L322](https://github.com/SonarSource/sonarqube/blob/eede4bb37ed5cbc255d60d3651ef969d44f6705f/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L322): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L344](https://github.com/SonarSource/sonarqube/blob/eede4bb37ed5cbc255d60d3651ef969d44f6705f/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L344): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDao.java#L372](https://github.com/SonarSource/sonarqube/blob/eede4bb37ed5cbc255d60d3651ef969d44f6705f/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDao.java#L372): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDto.java#L58](https://github.com/SonarSource/sonarqube/blob/eede4bb37ed5cbc255d60d3651ef969d44f6705f/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDto.java#L58): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. INFO [PermissionTemplateDto.java#L66](https://github.com/SonarSource/sonarqube/blob/eede4bb37ed5cbc255d60d3651ef969d44f6705f/sonar-db/src/main/java/org/sonar/db/permission/template/PermissionTemplateDto.java#L66): Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Pull request closed without merging**